pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go

```diff
@@ -0,0 +1,490 @@
+	// The flow for Update is the same as Add.
+	// A volume is only provisioned if not done so already.
+	controller.handleAddVolume(newObj)
+}
+
+func (controller *PersistentVolumeProvisionerController) handleAddClaim(obj interface{}) {
+	controller.mutex.Lock()
+	defer controller.mutex.Unlock()
+	cachedPvc, exists, _ := controller.claimStore.Get(obj)
+	if !exists {
+		glog.Errorf("PersistentVolumeClaim does not exist in the local cache: %+v", obj)
+		return
+	}
+	if pvc, ok := cachedPvc.(*api.PersistentVolumeClaim); ok {
+		err := controller.reconcileClaim(pvc)
+		if err != nil {
+			glog.Errorf("Error encoutered reconciling claim %s: %+v", pvc.Name, err)
+		}
+	}
+}
+
+func (controller *PersistentVolumeProvisionerController) handleUpdateClaim(oldObj, newObj interface{}) {
+	// The flow for Update is the same as Add.
+	// A volume is only provisioned for a claim if not done so already.
+	controller.handleAddClaim(newObj)
+}
+
+func (controller *PersistentVolumeProvisionerController) reconcileClaim(claim *api.PersistentVolumeClaim) error {
+	if controller.provisioner == nil {
+		return fmt.Errorf("No provisioner configured for controller")
+	}
+
+	// no provisioning requested, return Pending. Claim may be pending indefinitely without a match.
+	if !keyExists(qosProvisioningKey, claim.Annotations) {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] no provisioning required", claim.Name)
+		return nil
+	}
+	if len(claim.Spec.VolumeName) != 0 {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] already bound. No provisioning required", claim.Name)
+		return nil
+	}
+	if isAnnotationMatch(pvProvisioningRequiredAnnotationKey, pvProvisioningCompletedAnnotationValue, claim.Annotations) {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] is already provisioned.", claim.Name)
+		return nil
+	}
+
+	glog.V(5).Infof("PersistentVolumeClaim[%s] provisioning", claim.Name)
+	provisioner, err := newProvisioner(controller.provisioner, claim)
+	if err != nil {
+		return fmt.Errorf("Unexpected error getting new provisioner for claim %s: %v\n", claim.Name, err)
+	}
+	newVolume, err := provisioner.NewPersistentVolumeTemplate()
+	if err != nil {
+		return fmt.Errorf("Unexpected error getting new volume template for claim %s: %v\n", claim.Name, err)
+	}
+
+	claimRef, err := api.GetReference(claim)
+	if err != nil {
+		return fmt.Errorf("Unexpected error getting claim reference for %s: %v\n", claim.Name, err)
+	}
+
+	storageClass, _ := claim.Annotations[qosProvisioningKey]
+
+	// the creation of this volume is the bind to the claim.
+	// The claim will match the volume during the next sync period when the volume is in the local cache
+	newVolume.Spec.ClaimRef = claimRef
+	newVolume.Annotations[pvProvisioningRequiredAnnotationKey] = "true"
+	newVolume.Annotations[qosProvisioningKey] = storageClass
+	newVolume, err = controller.client.CreatePersistentVolume(newVolume)
+	glog.V(5).Infof("PersistentVolume[%s] created for PVC[%s]", newVolume.Name, claim.Name)
```

Comment:
To make sure I understand the flow:

When a new PVC object is created, `handleAddClaim(...)` is called which calls out to `reconcileClaim(...)` which creates a new PV object, if it decides to provision. At this point the provisioning hasn't actually happened yet, right (despite the log statement)? So this code sets the PV annotation `pvProvisioningRequiredAnnotationKey` to `true` and sets the PVC annotation `pvProvisioningRequiredAnnotationKey` to `pvProvisioningCompletedAnnotationValue`. Doing this ensures that `reconcileClaim(...)` will ignore the PVC it in the future.

The new PV object triggers `reconcileVolume(...)` which notices that the PV annotation `pvProvisioningRequiredAnnotationKey` is not `pvProvisioningCompletedAnnotationValue` and therefore triggers actual provisioning for the volume `provisioner.Provision(...)`